File: src/Services/ConferenceManagement/PaderConference.Core/Services/Synchronization/SynchronizedObjectId.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace PaderConference.Core.Services.Synchronization
{
    public readonly struct SynchronizedObjectId
    {
        public SynchronizedObjectId(string id, IReadOnlyDictionary<string, string> parameters)
        {
            Id = id;
            Parameters = parameters;
        }

        public SynchronizedObjectId(string id)
        {
            Id = id;
            Parameters = ImmutableDictionary<string, string>.Empty;
        }

        public string Id { get; }

        public IReadOnlyDictionary<string, string> Parameters { get; }

        public static SynchronizedObjectId Parse(string s)
        {
            var split = s.Split('?', 2);
            var id = split[0];
            var query = split.Length > 1 ? split[1] : string.Empty;

            var parameters = ParseQueryString(query);
            return new SynchronizedObjectId(id, parameters);
        }

        public override string ToString()
        {
            var text = $"{Id}";
            if (Parameters.Any()) text += "?" + BuildQueryString(Parameters);

            return text;
        }

        private static IReadOnlyDictionary<string, string> ParseQueryString(string s)
        {
            var parts = s.Split('&', StringSplitOptions.RemoveEmptyEntries).Select(x => x.Split('=', 2));
            return parts.ToDictionary(keyVal => keyVal[0], keyVal => keyVal[1]);
        }

        private static string BuildQueryString(IReadOnlyDictionary<string, string> parameters)
        {
            return string.Join('&', parameters.Select(x => $"{x.Key}={x.Value}"));
        }
    }
}

File: src/Web/WebSPA/Client/src/store/signal/synchronization/synchronized-object-id.ts
export function parseSynchronizedObjectId(s: string) {
   const split = s.split('?', 2);
   const id = split[0];
   const query = split.length > 1 ? split[1] : '';

   const parameters = parseQueryString(query);
   return { id, parameters };
}

export function synchronizedObjectIdToString(syncObjId: SynchronizedObjectId): string {
   let result = syncObjId.id;

   const parameters = Object.entries(syncObjId.parameters);
   if (parameters.length > 0) {
      result += '?' + buildQueryString(syncObjId.parameters);
   }

   return result;
}

function parseQueryString(s: string): { [key: string]: string } {
   const parts = s
      .split('&')
      .filter((x) => !!x)
      .map((x) => x.split('='));
   return Object.fromEntries(parts);
}

function buildQueryString(parameters: { [key: string]: string }) {
   return Object.entries(parameters)
      .map((x) => `${x[0]}=${x[1]}`)
      .join('&');
}

export type SynchronizedObjectId = { id: string; parameters: { [key: string]: string } };
